File: Leetcode July Challenge/Python/Max Sum of Rectangle No Larger Than K.py
import bisect
class Solution:
    def maxSumSubmatrix(self, matrix: List[List[int]], k: int) -> int:
        ret = -math.inf
        m, n = len(matrix), len(matrix[0])
        # make sure rows > columns, otherwise we should rotate the matrix by zip() function
        if m < n:
            matrix = list(zip(*matrix))
        m, n = max(m, n), min(m, n)
        for le in range(n):
            pre_sum = [0] * m
            for ri in range(le, n): 
                for r in range(m):
                    pre_sum[r] += matrix[r][ri]
                arr = [0, math.inf] 
                cur_sum = 0
                for x in pre_sum:
                    cur_sum += x
                    idx = bisect.bisect_left(arr, cur_sum - k)
                    ret = max(ret, cur_sum - arr[idx])
                    if ret == k: return k
                    bisect.insort(arr, cur_sum) 
        return ret

File: Leetcode July Challenge/C++/Max Sum of Rectangle No Larger Than K.cpp
class Solution {
public:
    

    int maxSumSubmatrix(vector<vector<int>>& m, int k) {
    int res = INT_MIN, rows = m.size(), cols = m[0].size();
    for (int l = 0; l < cols; ++l) {
        vector<int> sums(rows);
        for (int r = l; r < cols; ++r) {
            for (int i = 0; i < rows; ++i) 
                sums[i] += m[i][r];
            set<int> s = {0};
            int run_sum = 0;
            for (int sum : sums) {
                run_sum += sum;
                auto it = s.lower_bound(run_sum - k);
                if (it != end(s))
                    res = max(res, run_sum - *it);
                s.insert(run_sum);
            }
        }
    }
    return res;
}  
};
